### PR TITLE
Fix DEVSUP-799

### PR DESCRIPTION
**Pull request by jsteemann**
### Scope & Purpose

devel port of https://github.com/arangodb/arangodb/pull/14395

* Fixed DEVSUP-799: unique vertex getter may point to invalid memory after
  being resetted, resulting in undefined behavior for traversals returning
  unique vertices from inner FOR loops.

- [x] :hankey: Bugfix (requires CHANGELOG entry)
- [ ] :pizza: New feature (requires CHANGELOG entry, feature documentation and release notes)
- [ ] :fire: Performance improvement
- [ ] :hammer: Refactoring/simplification
- [x] :book: CHANGELOG entry made

#### Backports:

- [x] Backports required for: *3.7*: https://github.com/arangodb/arangodb/pull/14395, *3.8*: https://github.com/arangodb/arangodb/pull/14400, *3.6*: https://github.com/arangodb/arangodb/pull/14402

#### Related Information

- [x] GitHub issue / Jira ticket number: https://arangodb.atlassian.net/browse/DEVSUP-799

### Testing & Verification

- [x] The behavior in this PR was *manually tested*
- [x] This PR adds tests that were used to verify all changes:
  - [x] Added new **integration tests** (shell_server_aql)
- [x] I ensured this code runs with ASan / TSan or other static verification tools

### CODE REVIEW

File: arangod/Graph/Traverser.h [hunk 1]
```diff
@@ -158,6 +164,12 @@ class Traverser {
 
     void reset(arangodb::velocypack::StringRef const&) override;
 
+    void clear() override;
+
+#ifdef ARANGODB_ENABLE_MAINTAINER_MODE
+    bool pointsIntoTraverserCache() const noexcept override;
```
Comment (mchacki) [inline arangod/Graph/Traverser.h:170]: I like this.
This will help to track down those issues!

**Review by mchacki — APPROVED** [review]
LGTM.

Good Catch.
Thanks for the Fix.
No actions on my Comments required

**Comment by KVS85** [comment]
https://jenkins.arangodb.biz/job/arangodb-ANY-windows-packages/1729/

**Comment by KVS85** [comment]
> https://jenkins.arangodb.biz/job/arangodb-ANY-windows-packages/1742/

Crash wasn't reproduced.

**Pull request merged**